File: 2471-minimum-number-of-operations-to-sort-a-binary-tree-by-level/2471-minimum-number-of-operations-to-sort-a-binary-tree-by-level.java
import java.util.*;

class Solution {
    public int minimumOperations(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        int swp = 0;
        
        while (!q.isEmpty()) {
            int n = q.size();
            int[] v = new int[n];
            
            for (int i = 0; i < n; i++) {
                TreeNode cur = q.poll();
                v[i] = cur.val;
                if (cur.left != null) {
                    q.add(cur.left);
                }
                if (cur.right != null) {
                    q.add(cur.right);
                }
            }
            
            int[] sorted = v.clone();
            Arrays.sort(sorted);
            Map<Integer, Integer> idxMap = new HashMap<>();
            for (int i = 0; i < n; i++) {
                idxMap.put(v[i], i);
            }
            
            for (int i = 0; i < n; i++) {
                if (v[i] != sorted[i]) {
                    int idx = idxMap.get(sorted[i]);
                    idxMap.put(v[i], idx);
                    int temp = v[i];
                    v[i] = v[idx];
                    v[idx] = temp;
                    swp++;
                }
            }
        }
        
        return swp;
    }
}

File: 2471-minimum-number-of-operations-to-sort-a-binary-tree-by-level/2471-minimum-number-of-operations-to-sort-a-binary-tree-by-level.js
var minimumOperations = function(root) {
    const q = [root];
    let swp = 0;

    while (q.length) {
        const n = q.length;
        const v = [];
        
        for (let i = 0; i < n; i++) {
            const cur = q.shift();
            v.push(cur.val);
            if (cur.left) q.push(cur.left);
            if (cur.right) q.push(cur.right);
        }
        
        const sorted = [...v].sort((a, b) => a - b);
        const idxMap = new Map();
        for (let i = 0; i < n; i++) {
            idxMap.set(v[i], i);
        }
        
        for (let i = 0; i < n; i++) {
            if (v[i] !== sorted[i]) {
                const idx = idxMap.get(sorted[i]);
                idxMap.set(v[i], idx);
                [v[i], v[idx]] = [v[idx], v[i]];
                swp++;
            }
        }
    }
    
    return swp;
};
